src/query_graph/output.rs

```diff
@@ -0,0 +1,149 @@
+// Output module for query graphs
+// - Corresponds to the `graphviz` and `mermaid` modules from the JS federation.
+
+use crate::query_graph::{QueryGraph, QueryGraphEdge, QueryGraphNode};
+use apollo_compiler::NodeStr;
+use petgraph::dot::{Config, Dot};
+use petgraph::graph::{DiGraph, EdgeIndex};
+use petgraph::stable_graph::StableGraph;
+use std::fmt::Write;
+
+type InnerGraph = DiGraph<QueryGraphNode, QueryGraphEdge>;
+type StableInnerGraph = StableGraph<QueryGraphNode, QueryGraphEdge>;
+
+//////////////////////////////////////////////////////////////////////////////
+// GraphViz output for QueryGraph
+
+fn label_edge(edge: &QueryGraphEdge) -> String {
+    let label = edge.to_string();
+    if label.is_empty() {
+        String::new()
+    } else {
+        format!("label=\"{}\"", edge)
+    }
+}
+
+fn label_node(node: &QueryGraphNode) -> String {
+    format!("label=\"{}\"", node.type_)
+}
+
+pub fn to_dot(graph: &QueryGraph) -> String {
+    if graph.sources.len() > 1 {
+        return to_dot_federated(graph).expect("Failed to generate the federated graph");
+    }
+
+    // Note: Use label_edge/label_node as `attr_getters` in order to create custom label
+    //       strings, instead of the default labeling.
+    let config = [Config::NodeNoLabel, Config::EdgeNoLabel];
+    Dot::with_attr_getters(
+        &graph.graph,
+        &config,
+        &(|_, er| label_edge(er.weight())),
+        &(|_, (_, node)| label_node(node)),
+    )
+    .to_string()
+}
+
+fn to_dot_federated(graph: &QueryGraph) -> Result<String, std::fmt::Error> {
+    fn edge_within_cluster(
+        graph: &StableInnerGraph,
+        cluster_name: &NodeStr,
+        edge_index: EdgeIndex,
+    ) -> bool {
+        match graph.edge_endpoints(edge_index) {
+            None => false,
+
+            Some((n1, n2)) => {
+                graph[n1].source == *cluster_name && graph[n2].source == *cluster_name
+            }
+        }
```

Comment:
This is perfectly fine but you can also use the [`.is_some_and`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.is_some_and) method here:
```suggestion
        graph.edge_endpoints(edge_index).is_some_and(|(n1, n2)| {
            graph[n1].source == *cluster_name && graph[n2].source == *cluster_name
        })
```